core-api/src/main/java/com/optimizely/ab/Optimizely.java

```diff
@@ -923,9 +926,15 @@ public NotificationCenter getNotificationCenter() {
      * Convenience method for adding DecisionNotification Handlers
      */
     public int addDecisionNotificationHandler(NotificationHandler<DecisionNotification> handler) {
-        NotificationManager<DecisionNotification> manager = notificationCenter
-            .getNotificationManager(DecisionNotification.class);
-        return manager.addHandler(handler);
+        NotificationManager<DecisionNotification> notificationManager =
```

Comment:
Due to how Java implements generics, you cannot add a generic to the method signature. So to have a single method you'd need to do something like:
```
public <T> int addNotificationHander(Class<T> clazz, NotificationHandler<T> handler) {
        NotificationManager<T> notificationManager = notificationCenter.getNotificationManager(clazz);
        if (notificationManager == null) {
            logger.warn("Notification for class {}.", clazz);
            return -1;
        }

        return notificationManager.addHandler(handler);
    }
```
The only problem I found was that this was not discoverable. In a previous PR I got rid of `Notfification` as an interface which would have led to:
```
public <T> int addNotificationHander(Class<T extends Notification> clazz, NotificationHandler<T> handler) {
        NotificationManager<T> notificationManager = notificationCenter.getNotificationManager(clazz);
        if (notificationManager == null) {
            logger.warn("Notification for class {}.", clazz);
            return -1;
        }

        return notificationManager.addHandler(handler);
    }
```

Which would add to the discoverability, but I got some push-back to that interface. Thinking again perhaps it's best to reintroduce that.

Let me open another PR with that addition (although it will technically break 3.1 😬)

Reply:
@aliabbasrizvi See https://github.com/optimizely/java-sdk/pull/299 for what this would look like.

Reply:
After in person discussion, we'll keep the convenience methods but route to a generic impl.